### PR TITLE
Blog app - integration specs for views and fix n+1 problems

### PR DESCRIPTION
**Pull request by egichuhis**
Hello, please review this PR.

Changes:

1. Secure app from n+1 problems.

The N+1 problem was already solved when fetching all posts and their comments for a user.
Code snippet Posts controller:

posts

Logs snippet when fetching posts (includes comments)

logs

2. Write integration tests with Capybara gem.

- User Index Page:
Displays username, profile picture, and post count for each user.
Redirects to user's show page on click.

- User Show Page:
Shows user's profile picture, username, bio, and post count.
Displays the first three posts.
Provides a button to view all posts.
Redirects to post show page and user post index page as expected.

- User Post Index Page:
Displays user's profile picture, username, and post count.
Shows post title, body, first comments, comment and like counts.
Includes pagination for posts.
Redirects to post show page on post click.

- Post Show Page:
Displays post title, author, comments, and likes.
Shows post body, commenter username, and comments.

### CODE REVIEW

**Review by TedLivist — COMMENTED** [review]
## STATUS: INVALID CODE REVIEW REQUEST 🛑 🛑

Hi @egichuhis and @Rifat8080,

While you made a great effort in this project, unfortunately, I cannot proceed to review your code.

## Invalid Code Review Request

### Is @Nelofarzabi assigned to this team? I cannot see any commits from them. Kindly ensure all team members contribute to the project, and list all team members on the README. Kindly share a screenshot of your dashboard to clarify @Nelofarzabi is not part of the team.

Your Code Review Request will be marked as invalid in your Dashboard, so please submit a new one once you are ready 🙏

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear. And please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

--------

_Invalid Code Review Request **does not count** into the code reviews limit._

**Comment by egichuhis** [comment]
Dear reviewer, kindly note that this collaboration team has only 2 members. One of our members left within the week.

File: spec/posts_show_spec.rb
Comment (skyv26) [inline spec/posts_show_spec.rb:1]: - [x] Well done @egichuhis you and your coding partner did a good job, but it seems like one of you forgot to implement the `capybara` testing for the `post` show functionality. Kindly add the test cases for your `post` show and make sure that your tests are passing before asking for a review.

image

**Review by skyv26 — CHANGES_REQUESTED** [review]
## Hi @egichuhis and @Rifat8080, 

Good job so far 👍🏻 
There are some issues that you still need to work on to go to the next project but you are almost there!

## STATUS: REQUIRE CHANGES 🔴 



## Highlights
- Good **PR** title ✅ 
- Descriptive *PR* description ✅ 
- Implemented the features and fixed `N+1` problem ✅ 

## Required Changes ♻️

- [ ] Please make sure that your test cases are passing, because I checked that your test cases are passing except for the `User` index template. On the other hand, as I mentioned below in the `first` inline comment you missed the `implementation` of the test for the `post` show endpoint. Kindly take a look again and fix the problem,

### Results
image

- [x] It seems like you did great work and did a lot of research, I noticed that you are using the `type:: system` for your `controller` testing, which will open the `chromium` tab to automate the test of your application. Which is nice but make sure that your `tests` are also passing with `type::feature`. Please put your controller specs into a `dedicated` folder to organize everything in a well-structured manner.

image

- [x] Please add your `pair-programming` partner details under the `author` section, which not only gives the proper credit for the contribution but also shows how you are a good team player. 

image



_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better._

#### Cheers, and Happy coding!👏👏👏

## ⚠️  WARNING ⚠️ 
**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

## Have any doubt ❓ 
Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag @skyv26 in your question so I can receive the notification. You can also follow me on the below platforms

<table>
<tr>
<td>

<a href="https://github.com/skyv26">
   @skyv26
</a>

</td>
<td>

<a href="https://www.linkedin.com/in/skyv2022/">
   linkedin-skyv2022
</a>

</td>
<td>

<a href="https://twitter.com/vrma_aakash">
   @vrma_aakash
</a>

</td>
<td>

<a href="https://www.instagram.com/skybrel/">
   @skybrel
</a>

</td>
<td>

<a href="https://www.frontendmentor.io/profile/skyv26">
   @skybrel
</a>

</td>
</tr>
</table>

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/requests/users_spec.rb [hunk 1]
```diff
@@ -3,11 +3,10 @@
 RSpec.describe 'Users', type: :request do
   describe 'GET /users' do
     it 'renders the index template' do
-      get users_index_path
-      expect(response).to have_http_status(200)
+      get users_path
       expect(response).to render_template(:index)
```
Comment (tugsem) [inline spec/requests/users_spec.rb]: - [ ] Please fix this error to make all the tests pass. 
image

**Review by tugsem — CHANGES_REQUESTED** [review]
## STATUS: REQUIRED CHANGES ♻️
Hello @egichuhis 👋 , 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are some issues that you still need to work on to go to the next project but you are almost there!

Alt Text

 ## Highlights 🚀
✔️Good job on fixing the failing tests and adding tests for the post feature.

 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/requests/users_spec.rb (hunk 1 above)
Comment (egichuhis) [reply spec/requests/users_spec.rb]: All tests are passing. I have fixed and tested several times

tests_working

File: README.md [hunk 2]
```diff
@@ -89,6 +89,12 @@ Start your interactive ruby shell `irb`:
 - Twitter: [@twitterhandle](https://twitter.com/egichuhis)
 - LinkedIn: [LinkedIn](https://www.linkedin.com/in/edwin-gichuhi/)
 
+👤 **Md Mahadi Hasan Rifat**
+
+- GitHub: [@githubhandle](https://github.com/Rifat8080)
+- Twitter: [@twitterhandle](https://twitter.com/mdmahadirifat98)
+- LinkedIn: [LinkedIn](https://www.linkedin.com/in/md-mahadi-hasan-rifat-8b7851265/)
+
```
Comment (souad988) [inline README.md:97]: well done by adding your collaborators to the readme 👍🏼

File: spec/requests/users_spec.rb (hunk 1 above)
Comment (souad988) [reply spec/requests/users_spec.rb]: Good job @egichuhis the tests in the system folder are all running but still the first test in requests/user_spec file is failing please check your routes as it is triggering the index action in posts controller refer to the screenshot
Screenshot from 2023-12-15 14-49-56

**Review by souad988 — CHANGES_REQUESTED** [review]
Hi @egichuhis  

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.


 ## Highlights
Great efforts 
Good file structure
Communicate with reviewers using comments

 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by egichuhis** [comment]
Screenshot from 2023-12-16 08-41-44

**Review by mckent05 — APPROVED** [review]
Hi @egichuhis,

# STATUS: APPROVED ✔️ ✔️ 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations and well done! 🎉 
approved


## Highlights
- All linter checks are passing ✔️ ✔️ 
- Your PR is professional 👏 👏 

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**